File: pset6/credit/credit.py
from cs50 import get_string

#CheckMethod
def checkValidNumber(number):
    numberDigitCount = len(number)

    if ((numberDigitCount == 13) or (numberDigitCount == 15) or (numberDigitCount == 16)):
        sum = 0

        for i in range(1, numberDigitCount + 1, 1):
            temp = (int(number) % (pow(10,i)))

            if i > 1:
                temp = ((int(number) % (pow(10,i))) / (pow(10,i - 1)))

            if (i % 2 == 1):
                sum += temp
            else:
                sum += (temp * 2) / 10
                sum += (temp * 2) % 10

        if ((sum % 10) == 0):

            return False
        else:

            return True
    #Wrong digit length
    return False
#=====
def getCardSupplier(number):
    # numberDigitCount = len(number)

    # temp = int(number) / (pow(10 , (numberDigitCount - 2)))

    firstInt = number[0]

    secondInt = number[1]

    if (int(firstInt) == 4):
        return "VISA"
    if (int(firstInt) * 10 + int(secondInt)) in [34,37]:
        return "All American Express"
    if (int(firstInt) * 10 + int(secondInt)) in [51,52,53,54,55]:
        return "MasterCard"

    return "Invalid"
#=====
number = get_string("Number: ")

if checkValidNumber(number):
    cardSupplier = getCardSupplier(number)
    print(cardSupplier)

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>
int checkValidNumber(long long number);
char* cardSupplier(long long number);

int main(void)
{
    long long number;

    do {
        printf("Number: ");
        number = get_long_long("");
    }
    while (checkValidNumber(number) != 0) ;
    {
        printf("%s\n", cardSupplier(number));
    }
}

int checkValidNumber(long long number)
{
    int numberDigitCount = ceil(log10((double) number));
    //Check digit of number 
    if ((numberDigitCount == 13) || (numberDigitCount == 15) || (numberDigitCount == 16))
    {
        int sum = 0;
        
        for (int i = 1 ; i <= numberDigitCount ; i++)
        {
            int temp = (number % ((long long)pow(10,i))) ;
            
            if (i > 1)
            {
                temp = ((number % ((long long)pow(10,i))) / ((long long)pow(10,i - 1)));
            }
            
            if (i % 2 == 1) 
            {
                sum += temp ;
            }
            else 
            {
                sum += (temp * 2) / 10;
                sum += (temp * 2) % 10;
            }
        }
        if ((sum % 10) == 0) 
        {
            return 0;
        }
        else 
        {
            return 1;
        }
    }
    //Wrong digit length
    return 1;
}

char* cardSupplier(long long number) 
{
    int numberDigitCount = ceil(log10((double) number));
    
    int temp = number / ((long long) pow(10 , (numberDigitCount - 2)));
    
    int firstInt = temp / 10;
    
    int secondInt = temp % 10;
    
    
    if (firstInt == 4) 
    {
        return "VISA";
    } 
    else if ((firstInt == 3) && ((secondInt == 4) || (secondInt == 7)))
    {
        return "All American Express";
    }
    else if ((firstInt == 5) && ((secondInt == 1) || (secondInt == 2) || (secondInt == 3) || (secondInt == 4) || (secondInt == 5)))
    {
        return "MasterCard";
    }
    return "Invalid";
}